src/compile.rs

```diff
@@ -236,52 +387,118 @@ impl Compiler {
             Ok(())
         }
     }
+}
+
+#[derive(Debug)]
+enum Hole {
+    None,
+    One(InstIdx),
+    Many(Vec<Hole>),
+}
+
+#[derive(Clone, Debug)]
+enum MaybeInst {
+    Compiled(Inst),
+    Uncompiled(InstHole),
```

Comment:
Very nice observation. Thank you so much for taking the time to read this code enough to make this kind of suggestion! I've followed your advice and I agree the code is quite a bit clearer.